src/lib/http-observer.js

```diff
@@ -0,0 +1,223 @@
+        this.getAll(tabId),
+        this.getTypes(tabId)
+      );
+      logger.log(requestData);
+      return requestData;
+    }
+    logger.log('No requests available.');
+    return null;
+  }
+
+  // Passed a tab ID, returns an object with a key `all`, mapped to stats for
+  // all requests made in that tab.
+  getAll(tabId) {
+    return { all: this.getByType(tabId) };
+  }
+
+  // Passed a tab ID, returns an object mapping type => stats about the
+  // requests of that type for that tab.
+  getTypes(tabId) {
+    return REQUEST_TYPES.reduce(
+      (byType, type) => {
+        byType[type] = this.getByType(tabId, type);
+        return byType;
+      },
+      {}
+    );
+  }
+
+  // Passed a tabId and optional type, returns an object with stats for
+  // requests, filtered by type if one was provided.
+  getByType(tabId, type = null) {
+    const requests = this.filter(tabId, type);
+    return {
+      num: this.getNum(requests),
+      time: this.getTime(requests),
+      cached: this.getCached(requests),
+      cdn: this.getCDN(requests)
+    };
+  }
+
+  // Passed a tabId and optional type, returns the requests for that tab,
+  // filtered by type if one was provided.
+  filter(tabId, type = null) {
+    const requestIds = Object.keys(this.requests[tabId]);
+    return requestIds.reduce(
+      (filteredRequests, requestId) => {
+        const request = this.requests[tabId][requestId];
+        if (request.completed && (!type || type && request.type === type)) {
+          filteredRequests[requestId] = request;
+        }
+        return filteredRequests;
+      },
+      {}
+    );
+  }
+
+  // Passed an object mapping requestId => request, returns the number of
+  // requests.
+  getNum(requests) {
+    return Object.keys(requests).length;
+  }
+
+  // Passed an object mapping requestId => request, returns the time spent
+  // loading those requests.
+  getTime(requests) {
+    return Object.keys(requests).reduce(
+      (summed, requestId) => {
+        return summed + requests[requestId].elapsed;
+      },
+      0
+    );
+  }
+
+  // Passed an object mapping requestId => request, returns the proportion of
+  // those requests that were retrieved from cache. If there are no requests,
+  // returns null.
+  getCached(requests) {
+    const count = this.getNum(requests);
+    if (count === 0) {
+      return null;
+    }
+    return Object.keys(requests).reduce(
+      (cacheCount, requestId) => {
+        return cacheCount + requests[requestId].fromCache ? 1 : 0;
+      },
+      0
+    ) / count;
+  }
+
+  // Passed an object mapping requestId => request, returns the proportion of
+  // those requests that were retrieved from a CDN. If there are no requests,
+  // returns null.
+  getCDN(requests) {
+    const count = this.getNum(requests);
+    if (count === 0) {
+      return null;
+    }
+    return Object.keys(requests).reduce(
+      (cdnCount, requestId) => {
+        return cdnCount + requests[requestId].cdn ? 1 : 0;
```

Comment:
Not sure if **prettier** auto-formatted this, or if more readable if we just do a one-liner return function above.